### ISSUE TITLE
Another segfault, minimal repro to `zig run` / `zig test` below

### ISSUE THREAD

**Issue by metaleap** [ziglang/zig#4368]
`zig version`: `0.5.0+e6a812c82` (Zig's prebuilt-binary distr., approx. 1-2 days old), Void Linux x86_64, kernel 4.19.98_1

You'll encounter  `Segmentation fault` if you `zig build-exe` or `zig run` or `zig test` the following `minimal-repro.zig`.

Below the snippet: minor tweaks discovered to prevent the segfault, which allows building / running / testing without issue.

```zig
const std = @import("std");

test "" {
    main();
}

pub fn main() void {
    incomingRepro(@intCast(usize, std.time.milliTimestamp()));
}

pub fn incomingRepro(foo: usize) void {
    var msg: struct {
        result_ok: ?*std.json.Value = null,
        result_err: ?ResponseError = null,
    } = .{};

    inline for (@typeInfo(RequestOut).Union.fields) |*spec_field, idx| {
        if (foo == idx) {
            const TResponse = @typeInfo(@typeInfo(spec_field.field_type).Fn.return_type.?).Union.fields[0].field_type;
            const fn_arg = if (msg.result_err) |err|
                Ret(TResponse){ .err = err }
            else if (msg.result_ok) |ret|
                Ret(TResponse){ .ok = undefined }
            else
                unreachable;
        }
    }
}

pub const ResponseError = struct {
    code: isize,
    message: []const u8,
};

pub const RequestOut = union(enum) {
    pow2: fn (Arg(i64)) Ret(i64),
    rnd: fn (Arg(void)) Ret(f32),
    add: fn (Arg(?struct {
        a: i64,
        b: i64,
    })) Ret(?i64),
};

pub fn Arg(comptime T: type) type {
    return struct {
        it: T,
    };
}

pub fn Ret(comptime T: type) type {
    return union(enum) {
        ok: T,
        err: ResponseError,
    };
}
```

Preventing:

- **either:** change `unreachable` (L25) (the actually intended "value" here) to `undefined` or some `Ret(TResponse){...}` construction.
- **or:** remove/comment-out _any 2_ `RequestOut` union members (L36-38) so that the `union` has 1 member only (_any_ 1 of the 3 will do)

> Standard repro disclaimer: during minimal-repro-ification, ditching so much real-world surrounding stuff makes the above nonsensical and non-idiomatic. Ie. the now-vastly-reduced `msg` could be a union, `const Arg` now entirely superfluous, etc. etc. No matter, if the above is technically-valid Zig source, we don't want a segfault, and if it isn't, we also don't want a segfault  :smirk: 

Of note: if you add a type to `const fn_arg = if...` (L20) so that it's now `const fn_arg: Ret(TResponse) = if...`, the segfault also goes away, but the type-checker error message that then appears seems considerably nonsensical in the context above, and it is easily prevented by exactly the same 2 possible workarounds outlined above. Of which the first is more notable for the problem at hand I suspect, since the second merely turns the `inline for` into a just-once equivalent block (which could be telling, for troubleshooting --- or not).

**Comment by LemonBoy** [comment]
It's yet another freaky bug that has everything to do with `inline for` and the way it royally fucks up the basic-block contained inside: this time a `OptionalThen` is never analyzed at all and its `llvm_block` field is never set to a non-null value.

**Comment by Vexu** [comment]
Fixed by 4618c41fa6ca70f06c7e65762d2f38d57b00818c
